### ISSUE TITLE
Clean Up Settings

### ISSUE THREAD

**Issue by leon-richardt** [Chatterino/chatterino2#1956]
# What should be added?
Devise a way to clean up and organize Chatterino's "Settings" section. I mean this issue to be place to collect ideas on how to go about this.

## Basic & Advanced Section
Divide settings into a "Basic" and "Advanced" section, shoving obscure and rarely used options into the "Advanced" section. Sections could be divided by having separate tabs, for example.

For example, Chromium uses this approach:
<details>
<summary>Screenshot</summary>

![Chromium settings](https://user-images.githubusercontent.com/35232120/92950063-6496e280-f45c-11ea-8336-92fceb4bf8b7.png)
</details>

## "Developer Options"-style Settings
(Originally proposed by @Supinic.)

In order to access these settings, users would have to check a box first. Somewhat similar to the first approach, however it hides more obscure settings behind an additional barrier.

For example, Android uses this approach:
<details>
<summary>Screenshot</summary>

![Android settings](https://user-images.githubusercontent.com/35232120/92950620-45e51b80-f45d-11ea-91b5-35957756223b.png)
</details>

## Hide obscure settings altogether
This is the most radical approach but I will mention it anyway as it might inspire some other solution.

Remove obscure options from the GUI altogether and expose them only via editing the configuration file directly. This is probably a hassle to migrate users to though. (And thus very unlikely to be implemented.)

# Why should it be added?
Chatterino's settings are getting increasingly cluttered as the application grows in functionality. A search bar exists to quickly navigate to a setting but still, discoverability of features is hurt by having to scroll through many different options.

At the same time, the average user likely only cares about a small subset of settings available to them. To demonstrate, I compiled a quick list of configuration options that I think are irrelevant to most users:

<details>
<summary>List of rarely used settings</summary>

- "Show which channels are live in tabs" 
- "Show input when it's empty"
- "Show message length while typing"
- "Timestamp format"
- The "R9K" section
- The "Cache" section
- The "AppData" section
- "Restart on crash"
- "Use libsecret/KWallet/Gnome keychain to secure passwords"
- "Automatically close user popup when it loses focus"
- "Enable experimental IRC support"
- "Show unhandled IRC messages"

(This is not meant to be a complete list but rather to demonstrate that there are a lot of rarely changed settings.)
</details>

I am looking forward to other suggestions!

**Comment by fourtf** [comment]
I would like to see the "Basic & Advanced Section" added to the general settings.

**Comment by zneix** [comment]
How about adding some kind of guide through some settings for new / less advanced users to Chatterino's wiki / docs?

**Comment by leon-richardt** [comment]
> How about also adding some kind of guide through some settings for new / less advanced users to Chatterino's wiki / docs?

Definitely a good idea, although it is probably more approriate to discuss this over at [Chatterino/wiki](https://github.com/Chatterino/wiki).

**Comment by gempir** [comment]
What are you basic the list of rarely used settings on? 

I feel like without user tracking you can't just assume options aren't used. I personally use atleast 4 of the options.

**Comment by pajlada** [comment]
:+1: to Basic & Advanced Section

**Comment by leon-richardt** [comment]
> I feel like without user tracking you can't just assume options aren't used. I personally use atleast 4 of the options and I have heard of other users using some of the options.

Yeah, that will probably be the most tricky part of the change. I'm not sure what the best way to classify settings into "Basic" and "Advanced" without user-tracking would be.

Maybe some kind of poll among users? Feels a little "invasive" to me. And answers would probably skewed towards dedicated/power users anyway. :thinking:

**Comment by dnsge** [comment]
Idea:

For message options, what if the settings menu had a message preview that updated as you changed the settings? 
For example, a few messages in a `ChannelView` right next to settings like "Visible Badges", "Emotes", "Separate with lines",  "Alternate background color", "Timestamp format", etc.

**Comment by jupjohn** [comment]
I'd like to expand on some of the previous comments with a slightly modified layout:

 - Basic settings: anything that most people would probably care about (e.g. show badges, 
 - Advanced settings: slightly less used, but probably should still be changeable in-app
 - Settings file-only: settings that are intended primarily for developers/"power users" (i.e. audio backend, chat send protocol, most of the time gates)

I don't think we have a file watcher for settings in place, but if this approach was used I'd recommend adding one and potentially a popup to state that settings have changed.

**Comment by Mm2PL** [comment]
> Settings file-only: settings that are intended primarily for developers/"power users" (i.e. audio backend, chat send protocol, most of the time gates)
I'm highly against having more settings file only settings. Editing them is a nightmare. IMO only settings that should get a treatment like that are time gates for which we have helix and no other option. And even these have problems: they have had GUI controls, we probably should expect users to change them, taking away that GUI makes it way harder to reset them or debug behaviour caused by them.

**Comment by jupjohn** [comment]
@Mm2PL that's a fair point around removing controls that users may "depend on;" I guess I'm only talking about super niche settings that regular users wouldn't want to change (which is subjective, eh.)

With neri's tinegate cleanup I can only see a few items that I'd class as meaningless to the average user. Perhaps a "developers" subsection of advanced would suffice.

**Comment by jupjohn** [comment]
Here's a full-size screenshot of the entire settings window as of 3e5b4a9d0 (I needed this):

![chatterino settings window with all settings listed](https://github.com/user-attachments/assets/9dd004ff-d038-469a-aab3-5600797bf4bc)

**Comment by jupjohn** [comment]
Spent a bit of time thinking about ways we could incrementally tackle this (because sorting this all out in one go sounds like too much) and came up with the following:

1. Split out the top level categories in `General` to "sub-pages" (listed on the left) of a _General_/_Basic_
    1. Splitting them out makes it a lot easier to parse through - for both this cleanup and users;
    2. It also makes it easier for users to find things as most people (from what I'm aware of, could poll it) don't scan from the left bar, OVER the settings themselves, to the _ToC_ pane on the right - they'll go straight into the middle settings page;
    3. You'd end up removing the right-hand _ToC_ pane, which gives more horizontal real-estate to work with
2. Move _Advanced_ settings to their own top-level page with categories (like proposed above)
3. We start tagging all settings with a bunch of key words (i.e. "_layout_," "_font_," "_user_")
	1. This should make categorising settings into new sub-pages/sections a lot easier
	2. Tagging either in code (comments/some kind of list on the setting instance) or in some shared place would mean we can spread the load of this across people who want to contribute
	3. Tags could be searchable, making search results a bit richer (or worse if there are too many tags)
4. Once things get tagged, settings can be re-categorised based on them
	1. Might as well tackle the ordering of most-useful to least-useful (per-category) at the same time
5. As some pages have tabs to jump between different sub-pages, changing those tabs into sub-pages like _General_ & _Advanced_

Thoughts? Questions? I feel like at least changing the one massive _General_ page into sub-pages would be a good start.
